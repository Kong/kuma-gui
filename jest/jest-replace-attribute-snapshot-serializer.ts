// Imports the appropriate plugin type from the library Jest uses for snapshot serialization.
import { NewPlugin } from 'pretty-format'

const DEFAULT_UUID = 'aaaabbbb-cccc-dddd-eeee-ffffffffffff'

function hasUuidAttribute(element: Element, attribute: string): boolean {
  const value = element.getAttribute(attribute)

  return value !== null && /[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}/.test(value)
}

const processedValues = new Set()

export function replaceAttributesSnapshotSerializer(
  attributes: string[],
  uuid = DEFAULT_UUID,
): NewPlugin {
  return {
    /**
     * Matches elements which havenâ€™t been processed by this serializer before.
     */
    test: (value) => value instanceof Element && !processedValues.has(value)
      ? attributes.some((attribute) => hasUuidAttribute(value, attribute))
      : false,

    /**
     * Replaces the attribute values of an element with a static value.
     */
    serialize: (value: Element, config, indentation, depth, refs, printer) => {
      for (const attribute of attributes) {
        if (hasUuidAttribute(value, attribute)) {
          value.setAttribute(attribute, uuid)
        }
      }

      processedValues.add(value)

      return printer(value, config, indentation, depth, refs)
    },
  }
}
