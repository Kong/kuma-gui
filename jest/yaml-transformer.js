const transformer = require('yaml-jest').default

module.exports = {
  ...transformer,
  process: function (...params) {
    return {
      code: transformer.process(...params),
      map: null,
    }
  },
}
