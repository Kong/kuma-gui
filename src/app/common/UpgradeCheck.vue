<template>
  <DataSource
    v-slot="{ data }"
    :src="`/control-plane/version/latest`"
  >
    <KAlert
      v-if="data && env('KUMA_VERSION') !== data.version"
      data-testid="upgrade-check"
      class="upgrade-check-alert"
      appearance="info"
      size="small"
    >
      <template #alertMessage>
        <div class="alert-content">
          <div>
            {{ t('common.product.name') }} update available
          </div>

          <div>
            <KButton
              appearance="primary"
              :to="t('common.product.href.install')"
            >
              Update
            </KButton>
          </div>
        </div>
      </template>
    </KAlert>
  </DataSource>
</template>

<script lang="ts" setup>
import { useEnv, useI18n } from '@/utilities'
const env = useEnv()
const { t } = useI18n()
</script>

<style lang="scss" scoped>
.k-alert.small {
  // Uses smaller paddings for this particular alert.
  padding: $kui-space-20 $kui-space-40;
}

.alert-content {
  display: flex;
  align-items: center;
  font-size: $kui-font-size-30;

  > *:first-of-type {
    margin-right: $kui-space-50;
  }
}
</style>
