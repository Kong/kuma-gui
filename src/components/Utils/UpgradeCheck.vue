<template>
  <div class="upgrade-check">
    <KAlert
      v-if="showNotice"
      appearance="warning"
      size="small"
    >
      <template slot="alertMessage">
        <div class="alert-content">
          <div>
            {{ tagline }} update available
          </div>
          <div>
            <KButton
              class="warning-button"
              appearance="primary"
              size="small"
              :to="url"
            >
              Update
            </KButton>
          </div>
        </div>
      </template>
    </KAlert>
  </div>
</template>

<script>
import axios from 'axios'
import compare from 'semver-compare'
import { mapGetters } from 'vuex'

export default {
  name: 'UpgradeCheck',
  data () {
    return {
      url: `${process.env.VUE_APP_INSTALL_URL}${process.env.VUE_APP_UTM}`,
      latestVerSrc: process.env.VUE_APP_VERSION_URL,
      latestVer: null,
      showNotice: false
    }
  },
  computed: {
    ...mapGetters({
      currentVer: 'config/getVersion',
      tagline: 'config/getTagline'
    })
  },
  beforeMount () {
    this.checkVersion()
  },
  methods: {
    checkVersion () {
      axios.get(this.latestVerSrc)
        .then(response => {
          const status = response.status
          const data = response.data

          if (status === 200 && data && data.length > 0) {
            this.latestVer = data
          }
        })
        .catch(error => {
          this.showNotice = false

          console.error(error)
        })
        .finally(() => {
          if (this.latestVer) {
            // compare the latest version to the currently running version
            // but only if we were able to set the latest version in the first place.
            const comparison = compare(this.latestVer, this.currentVer)

            if (comparison === 1) {
              this.showNotice = true
            } else {
              this.showNotice = false
            }
          } else {
            const timespan = 3 // months
            const today = new Date()
            const refDate = new Date('2020-06-03 12:00:00')
            const later = new Date(
              refDate.getFullYear(),
              refDate.getMonth() + timespan,
              refDate.getDate()
            )

            // compare dates and handle the notice accordingly
            if (today.getTime() >= later.getTime()) {
              this.showNotice = true
            } else {
              this.showNotice = false
            }
          }
        })
    }
  }
}
</script>

<style lang="scss" scoped>
.upgrade-check {

  .k-alert.k-alert {
    padding: var(--spacing-xxs) var(--spacing-xs);
  }
}

.alert-content {
  display: flex;
  align-items: center;
  font-size: var(--type-sm);

  > *:first-of-type {
    margin-right: var(--spacing-sm);
  }

  .k-button {
    text-decoration: none;

    &:after {
      display: none;
    }
  }
}

.warning-button {
--KButtonPrimaryBase: #f2a230;
--KButtonPrimaryHover: #f2a230;
--KButtonPrimaryActive: #f2a230;
}

</style>
