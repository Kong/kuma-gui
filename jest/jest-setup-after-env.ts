import { afterAll, afterEach, beforeAll, beforeEach, expect, jest } from '@jest/globals'

// Polyfills `window.fetch` for Jest because it runs in a Node environment where fetch isn’t available. It initially looked like this would change with Node.js 18, but that is not so.
import 'isomorphic-fetch'

import { config } from '@vue/test-utils'

import { replaceAttributesSnapshotSerializer } from './jest-replace-attribute-snapshot-serializer'
import { createRouter } from '../src/router/router'
import { TOKENS, get, container, set, injected } from '../src/services'
import { TOKENS as COMPONENT_TOKENS } from '../src/components'
import { setupMockServer } from '../src/api/setupMockServer'
import { rest, MockedRequest as Request } from 'msw'
import Env from '@/services/env/Env'

type MockFunction = (_opts: Record<string, unknown>, cb: (req: Request, resp: Record <string, any>) => Record<string, unknown>) => void

// jest can't import this module properly due to transpiling issues
// mock this out with a blank element
jest.mock('vue-github-button', () => ({ template: '<span />' }))

/**
 * Adds the application’s router to vue test utils. This way tests don’t have to set-up a new router instance on their own.
 */
const router = createRouter(get(TOKENS.routes))
config.global.plugins.push(router)

/**
 * Adds the application’s Vuex store to vue test utils. This way tests don’t have to set-up a new store instance on their own.
 */
config.global.plugins.push([get(TOKENS.store), get(TOKENS.storeKey)])

/**
 * Kongponents uses generated UUIDs for several attribute values.
 * This breaks the project’s snapshot tests since they’re based on fully-mounted components
 * which also includes those from external sources like Kongponents.
 *
 * In order to stabilize the tests which otherwise fail because the attribute values are different every run,
 * we use a custom snapshot serializer to replace those attribute values with one fixed value.
 */
expect.addSnapshotSerializer(replaceAttributesSnapshotSerializer([
  'id',
  'aria-describedby',
  'aria-labelledby',
  'aria-controls',
  'data-tableid',
]))

const server = setupMockServer(import.meta.env.VITE_KUMA_API_SERVER_URL)

beforeAll(() => server.listen())
afterEach(() => server.resetHandlers())
afterAll(() => server.close())

// unless we actually use COMPONENT_TOKENS it won't actually get executed
// probably due to tree shaking/rollup import ordering. This mixed with
// container capturing/restoring to make our tests isolated means that
// potentially we can capture an empty container before all the tokens are set,
// then set the TOKENS/fill the container during a test then the container can
// get restored to empty whilst we still have TOKENS with now non-existent
// services accessing TOKENS before we do anything means we set the TOKENS and
// fill the container with the default services i.e. before we capture if we
// ever make a test mocking utility to mock out components (similar to
// withVersion) will will then use COMPONENT_TOKENS here also, which means we
// can remove the following line
beforeAll((_ = COMPONENT_TOKENS) => {})
//
beforeEach(() => container.capture?.())
afterEach(() => container.restore?.())

// add a utility to easily setup/mock out API endpoints
const re = /\+/g

const useMock = (url: string, response: Record<string, unknown>):MockFunction => {
  return (_opts, cb) => {
    server.use(
      rest.get(`${import.meta.env.VITE_KUMA_API_SERVER_URL.slice(0, -1)}${url.replace(re, '\\+')}`, (req, res, ctx) => {
        return res(ctx.json(cb(req, JSON.parse(JSON.stringify(response)))))
      }),
    )
  }
}

export const withVersion = (v: string) => {
  class TestEnv extends Env {
    var(...rest: Parameters<Env['var']>) {
      const key = rest[0]
      if (key === 'KUMA_VERSION') {
        return v
      }
      return super.var(...rest)
    }
  }
  set(TOKENS.Env, TestEnv)
  injected(TestEnv, TOKENS.EnvVars)
}
export { router, server, useMock }
