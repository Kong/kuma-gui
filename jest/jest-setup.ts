import '@testing-library/jest-dom/extend-expect'
import '@testing-library/jest-dom'

// Polyfills `window.fetch` for Jest because it runs in a Node environment where fetch isn’t available. It initially looked like this would change with Node.js 18, but that is not so.
import 'isomorphic-fetch'

import dotenv from 'dotenv'

import { replaceAttributesSnapshotSerializer } from './jest-replace-attribute-snapshot-serializer'
import { setupMockServer } from '../src/api/setupMockServer'

/**
 * Ensures the tests run with the project’s environment variables being available.
 */
dotenv.config()

/**
 * amcharts need SVGPathElement defined for our tests to work.
 *
 * See: https://github.com/amcharts/amcharts4/issues/1387
 */
Object.defineProperty(window, 'SVGPathElement', { value: class extends HTMLElement { } })

/**
 * Kongponents uses generated UUIDs for several attribute values.
 * This breaks the project’s snapshot tests since they’re based on fully-mounted components
 * which also includes those from external sources like Kongponents.
 *
 * In order to stabilize the tests which otherwise fail because the attribute values are different every run,
 * we use a custom snapshot serializer to replace those attribute values with one fixed value.
 */
expect.addSnapshotSerializer(replaceAttributesSnapshotSerializer([
  'id',
  'aria-describedby',
  'aria-labelledby',
  'aria-controls',
  'data-tableid',
]))

const server = setupMockServer(import.meta.env.VITE_KUMA_API_SERVER_URL)

// Establish API mocking before all tests.
beforeAll(() => server.listen())
// Reset any request handlers that we may add during the tests,
// so they don't affect other tests.
afterEach(() => server.resetHandlers())
// Clean up after the tests are finished.
afterAll(() => server.close())

export { server }
