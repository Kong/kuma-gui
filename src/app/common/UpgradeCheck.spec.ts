import { describe, expect, test } from '@jest/globals'
import { flushPromises, mount } from '@vue/test-utils'

import UpgradeCheck from './UpgradeCheck.vue'
import { withVersion } from '@/../jest/jest-setup-after-env'

function renderComponent() {
  return mount(UpgradeCheck)
}

describe('UpgradeCheck.vue', () => {
  test('renders snapshot', async () => {
    withVersion('1.2.0')

    const wrapper = renderComponent()

    await flushPromises()
    expect(wrapper.html()).toContain('Update')

    expect(wrapper.element).toMatchSnapshot()
  })
})
