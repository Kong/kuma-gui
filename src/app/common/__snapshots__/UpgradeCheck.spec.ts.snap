// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UpgradeCheck.vue renders snapshot 1`] = `
<div
  class="upgrade-check"
>
  <div
    class="warning small none k-alert upgrade-check-alert"
    role="alert"
  >
    <!---->
    <!---->
    <div
      class="k-alert-msg-text"
    >
      <!---->
      <div
        class="k-alert-msg"
      >
        
        <div
          class="alert-content"
        >
          <div>
            Kuma update available 
          </div>
          <div>
            <a
              class="small rounded primary k-button warning-button"
              href="https://kuma.io/install/latest/?utm_source=Kuma&utm_medium=Kuma-GUI"
              type="button"
            >
              
              <!---->
              
              
               Update 
              
              <!---->
            </a>
          </div>
        </div>
        
      </div>
      <!---->
    </div>
    <!---->
    <!---->
  </div>
</div>
`;
